packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue/queue.ts

```diff
@@ -0,0 +1,389 @@
+ */
+interface QueueAutoScalingOptions {
+  /**
+   * Acceptable amount of time taken to process one message in the queue.
+   */
+  readonly messageProcessingTime: cdk.Duration;
+
+  /**
+   * Acceptable amount of time for processing a single message in the queue.
+   */
+  readonly acceptableLatency: cdk.Duration;
+}
+
+/**
+ * The `TopicSubscription` class represents an SNS Topic resource that can be subscribed to by the service queues.
+ */
+export class TopicSubscription implements ISubscribable {
+  public readonly topic: sns.ITopic;
+
+  public readonly queue?: sqs.IQueue;
+
+  private readonly subscriptionQueue?: SubscriptionQueue;
+
+  constructor(props: TopicSubscriptionProps) {
+    this.topic = props.topic;
+    this.subscriptionQueue = props.topicSubscriptionQueue;
+    this.queue = props.topicSubscriptionQueue?.queue;
+  }
+
+  /**
+   * This method sets up SNS Topic subscriptions for the SQS queue provided by the user. If a `queue` is not provided,
+   * the default `eventsQueue` subscribes to the given topic.
+   *
+   * @param extension `QueueExtension` added to the service
+   * @returns the queue subscribed to the given topic
+   */
+  public subscribe(extension: QueueExtension) : SubscriptionQueue {
+    let ret = {
+      queue: extension.eventsQueue,
+      scaleOnLatency: extension.autoscalingOptions,
+    } as SubscriptionQueue;
+
+    if (this.subscriptionQueue) {
+      ret = this.subscriptionQueue;
+    }
+    this.topic.addSubscription(new subscription.SqsSubscription(ret.queue));
+    return ret;
+  }
+}
+
+/**
+ * Settings for the hook which mutates the application container
+ * to add the events queue URI to its environment.
+ */
+interface ContainerMutatingProps {
+  /**
+   * The events queue name and URI to be added to the container environment.
+   */
+  readonly environment: { [key: string]: string };
+}
+
+/**
+ * This hook modifies the application container's environment to
+ * add the queue URL for the events queue of the service.
+ */
+class QueueExtensionMutatingHook extends ContainerMutatingHook {
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default `eventsQueue` for the service (if not provided) and accepts a list of objects of
+ * type `ISubscribable` that the `eventsQueue` subscribes to. It creates the subscriptions and sets up permissions
+ * for the service to consume messages from the SQS Queues.
+ *
+ * It also configures auto scaling for the service by scaling up or down to maintain an acceptable queue latency by
+ * tracking the accpetable backlog per task.
```

Comment:
Spelling: 
```suggestion
 * tracking the acceptable backlog per task.
```

Reply:
Would it make sense to use the term "calculating" instead of "tracking"? I think "tracking" seems to imply that during runtime of the service, it will do some magic to scale the service based on the acceptable latency and the **actual** message processing time. But that is not the case right? The user must supply a value for the message processing time. Am I understanding the feature correctly?

Reply:
I think it should be `tracking the backlog per task`. We use the Lambda function to periodically emit this metric to CloudWatch Logs and use a target tracking policy to 'track' this value. I've updated the comment to make it clearer!